### ISSUE TITLE
Other than float64 results in `TypeError`

### ISSUE THREAD

**Issue by zerothi** [pyamg/pyamg#214]
When forcing the dtype to be `float32` I get typeerrors:

I can reproduce this with this small snippet:
```
from __future__ import print_function
import numpy as np
import pyamg

sh = (40, 40)
for dtype in [np.float64, np.float32]:
    A = pyamg.gallery.poisson(sh, dtype=dtype, format='csr')
    b = np.random.rand(A.shape[0]).astype(dtype)

    # Assert dtypes
    print(A.dtype, b.dtype)

    ml = pyamg.aggregation.smoothed_aggregation_solver(A)
    x = ml.solve(b)

    print("residual: ", np.linalg.norm(b-A*x))
```
The first run goes fine, the second produces a TypeError because somewhere down the line, the vectors are upcasted (or something is at least).

PYAMG: 3.3.2, scipy 1.0.0, numpy 1.13.3

**Comment by lukeolson** [comment]
Ok, I can reproduce the error and will dig into this...

**Comment by lukeolson** [comment]
Ok, fixed this in 7d5ec7b and added a test.

Small modification of your example (for repeatability):
```
from __future__ import print_function
import numpy as np
import pyamg

sh = (40, 40)
for dtype in [np.float64, np.float32]:
    np.random.seed(0)
    A = pyamg.gallery.poisson(sh, dtype=dtype, format='csr')
    b = np.random.rand(A.shape[0]).astype(dtype)

    # Assert dtypes
    print(A.dtype, b.dtype)

    ml = pyamg.aggregation.smoothed_aggregation_solver(A)
    x = np.random.rand(A.shape[0]).astype(dtype)
    x = ml.solve(b, x0=x, maxiter=2)

    print("residual: ", np.linalg.norm(b-A*x))
```

Thanks for reporting this bug!
